### ISSUE TITLE
rc16 seems quite broken

### ISSUE THREAD

**Issue by ralleman** [oracle/graal#1231]
ECMA import support is no longer working. This was working in rc14, but not in rc16:

```javascript
import('/compiler.mjs')
  .then(module => {
    let compiler = new module.Compiler(app);
    compiler.compile();

    api.applications.linkDomain(app.$id, domain);
    return true;
  })
  .catch(error => {
    print('Oh dear!');
    print(error);
    error.printStackTrace();
    throw error;
  });
```

rc16 gives the error:

```javascript
SyntaxError: /js/main.js:1:0 Expected an operand but found import
import('/compiler.mjs')
^
/js/main.js:2:2 Expected an operand but found .
  .then(module => {
  ^
/js/main.js:7:4 Invalid return statement
    return true;
    ^
/js/main.js:8:2 Expected eof but found }
  })
  ^
```

I've detailed another instance of something that was working in rc14, but fails in rc16 here https://github.com/oracle/graal/issues/1134.

Could the rc16 gz artifact have corrupt files in it?

**Comment by ralleman** [comment]
Turns out rc15 has the same fault above. I'm going to try rc13 because rc14 has the fault mentioned in #1134 which is now blocking my work.

**Comment by ralleman** [comment]
rc13 has the same problem described in #1134, so I'm now properly blocked. Can either this issue or #1134 be escalated please?

**Comment by iamstolis** [comment]
You are using dynamic import that is not part of the ECMAScript specification yet. GraalVM JavaScript supports dynamic import but its support is disabled by default. You can enable it by setting `js.ecmascript-version` to `11`.

**Comment by ralleman** [comment]
Thanks @iamstolis. That got me past that error and onto another error:

`TypeError: invokeMember on JavaObject[com.mydomain.script.ApiProxyModulesService$1@79aa675b (com.mydomain.script.ApiProxyModulesService$1)] failed due to: Unknown identifier: find`

Where the Java code is:

```java
@Service
public class ApiProxyModulesService {

    public interface ModulesProxy {
        @SuppressWarnings("unused")
        Value find(String name);

        @SuppressWarnings("unused")
        Value find(String name, String version);
    }

    @Autowired
    private ProxyService proxyService;

    @Autowired
    private InbuiltApplicationService inbuiltApplicationService;

    /**
     * Returns a new proxy object for inbuilt module loading.
     */
    public Object getModulesProxy(CallContext callContext, Context context) {
        return new ModulesProxy() {

            public Value find(String name) {
                if (name == null) {
                    throw new RuntimeException("Missing module name");
                }

                Resource resource = inbuiltApplicationService.getModule(name, null);
                if (resource == null) {
                    return null;
                }

                return proxyService.getProxyInstance(callContext, context, resource);
            }

            public Value find(String name, String version) {
                if (name == null) {
                    throw new RuntimeException("Missing module name");
                }

                Resource resource = inbuiltApplicationService.getModule(name, version);
                if (resource == null) {
                    return null;
                }

                return proxyService.getProxyInstance(callContext, context, resource);
            }
        };
    }
}
```

And the javascript is:

```javascript
        // Attempt to find the module from the common pool first.
        let module = api.modules.find(this.name, this.version);
        if (module != null) {
          return module;
        }
```

Again, this was all working with rc14. rc16 is having trouble invoking api.modules.find (api.modules is the object returned from getModulesProxy(...)). Any idea?

Is there any doco on the options anywhere?

**Comment by iamstolis** [comment]
There were some recent changes that made `Context` more restricted by default. I think that you are running into one of these restrictions (arbitrary methods/fields of Java classes are not accessible by default). You can remove these restrictions using `allowAllAccess(true)` method (or some more specific `allow*()` method) on `Context.Builder`.

**Comment by ralleman** [comment]
allowAllAccess(true) gets past it, but I can't use this because I run untrusted code. If this is the only way I can get past it, it's a huge step backwards on running untrusted code. Is there another way past this that doesn't require me to allow access to untrusted code?

Additionally, while allowAllAccess(true) gets me running(ish) I'm still blocked by #1134 except the error message has changed to "writeMember on foreign object failed due to: Message not supported" so rc16 hasn't gotten me past the main error.

**Comment by iamstolis** [comment]
>  Is there another way past this that doesn't require me to allow access to untrusted code?

`Context.Builder` has various methods that allow you to specify fine-grained privileges of the `Context`. I suggest you to start with an appropriate `HostAccess` (`allowHostAccess(HostAccess.newBuilder()....build())`).

>  I'm still blocked by #1134 ...

I am sorry, I cannot help you with #1134 as it does not contain a reproducible test-case. There is some JavaScript code but if I understand correctly the same JavaScript code works well with some of your `ProxyObject`s but not with other. So, the root of the problem is in the implementation of the problematic `ProxyObject` probably. I suggest you to use Java debugger to figure out the difference between the working and non-working `ProxyObject`s.

**Comment by ralleman** [comment]
> Context.Builder has various methods that allow you to specify fine-grained privileges of the Context. > I suggest you to start with an appropriate HostAccess (allowHostAccess(HostAccess.newBuilder()....build())).
Understood. That's working. Thanks.

Re #1134, I have not been able to reproduce this in an isolated test case, but I can confirm that I have hundreds of instances of a ProxyObject class called ResourceProxy and one of those instances is consistently producing the error. It's almost like it reaches a max count of some kind because putMember is not even getting invoked on that one instance (like it's saying, no, I'm invoked that method too many times already!). The ProxyObject is the same type as all the others and the input values are are same. If I could understand each of the valid reasons for the error, I maybe able to produce a test case for it.

**Comment by ralleman** [comment]
@iamstolis, I've resolved #1134. Thanks for your help here. Closing this.
